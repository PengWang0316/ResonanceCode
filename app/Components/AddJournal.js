import React, { Component } from "react";
import Util from "../apis/Util";
import Loading from "./Loading"
import JournalContent from "./JournalContent";
import ReadingSearchAndList from "./ReadingSearchAndList";
import { isLogin } from "../apis/LoginApi";
import DatabaseApi from "../apis/DatabaseApi";
import PropTypes from "prop-types";

class AddJournal extends Component {

  componentWillMount(){
    // Initial variables
    this.readings = this.props.journalData ? this.props.journalData.readingIds : {}; // keep which readings should be attached on. Format is like {readingId: null}
    this.contents = {}; // Keep content keys and content
    this.contentKeyIndex = 0; // Generate keys for different contents
    this.contentIndexs = {}; // Use to track the index of content component in the array. Delete function needs it. The format is {contentKey: index}
    this.jounalId = this.props.journalData ? this.props.journalData._id : null; // Keeping journal id for update
    this.userId = isLogin(document).userid;

    this.state={
      isWriting: false,
      journalDate: this.props.journalData ? Util.getDateString(this.props.journalData.date) : Util.getCurrentDateString(),
      isDateCorrect: true,
      isEmptyReading: this.props.journalData ? false : true,
      contentComponentArray: [], // keep content component
      addJournalContent: "overview",
      pingPongState: this.props.journalData ? this.props.journalData.ping_pong_state : "Neutral"
    };

    console.log("Journal before alter: ", this.props.journalData);
    // If journal data exsit, get content
    if(this.props.journalData){
      // this.oldContentKeys = []; // keep original content keys
      this.oldReadingIds = Object.keys(this.readings); // keep original reading ids

      // delete unnecessary properties from journal object
      let journal = this.props.journalData;
      this.journalUserId = journal.user_id;
      delete journal.date;
      delete journal.user_id;
      delete journal.ping_pong_state;
      delete journal._id;
      delete journal.readingIds;
      let keyRegExp = /-(\d+)$/; // The regular expression for subtract suffixs
      console.log("Journal after alter: ", this.props.journalData);
      Object.keys(journal).map((key)=>{
        // this.oldContentKeys.push(key); // keeping the exsit keys in an array for delete function
        // handleAddContentClick(addJournalContent, newContentName, newContentKey, contentKeyIndex, isShared)
        // Also have to save the largest key number and use it to countine
        // journalData property is like {'overview-0': 'test overview', 'overview-0-isShared': true}
        // if key is isShared do not do anything
        let matchResult = key.match(keyRegExp);
        if (matchResult){
          if (this.contentKeyIndex < matchResult[1]) this.contentKeyIndex = matchResult[1]; // keep the max index number and user can contiune add new content
          this.handleAddContentClick(journal[key], key.replace(keyRegExp, ""), key, matchResult[1], journal[`${key}-isShared`]);
          // console.log("isShared: ",journal[`${key}-isShared`]);
        }
      });
      this.contentKeyIndex++;
    }

  }

  handleChange(event, element){
    let newState={};
    newState[element]=event.target.value;
    if(element=="journalDate") newState.isDateCorrect = Util.matchDateFormat(newState[element]);
    this.setState(newState);
  }

  handleChangeCallback(contentKey, contentText){
    console.log(contentKey, contentText);
    // update content in state
    this.contents[contentKey]=contentText;
  }

  handleDeleteContentCallback(contentKey){
    // Remove content from state
    if(this.contents.hasOwnProperty(contentKey)){
      delete this.contents[contentKey];
      delete this.contents[`${contentKey}-isShared`];
    }
    // Remove content component
    delete this.contentIndexs[contentKey];
    this.setComponentToStateArray();
    // this.state.contentComponentArray.splice(this.contentIndexs[contentKey],1);
    // this.setState({contentComponentArray: this.state.contentComponentArray});
  }

  handleSharedBoxChangeCallback(contentKey, isShared){
    this.contents[`${contentKey}-isShared`] = isShared;
  }

  handleAddContentClick(addJournalContent, newContentName, newContentKey, contentKeyIndex, isShared){
    // put component in array in order to show
    newContentKey = newContentKey ? newContentKey : `${this.state.addJournalContent}-${this.contentKeyIndex}`;
    // put the component in the track object
    this.contentIndexs[newContentKey]=<JournalContent key={contentKeyIndex ? contentKeyIndex : this.contentKeyIndex++} newContent={addJournalContent ? addJournalContent : ""} newContentName={newContentName ? newContentName : this.state.addJournalContent} newContentKey={newContentKey} handleChangeCallback={(contentKey, contentText)=>{this.handleChangeCallback(contentKey, contentText);}} handleDeleteContentCallback={(contentKey)=>{this.handleDeleteContentCallback(contentKey);}} handleSharedBoxChangeCallback={(contentKey, isShared)=>{this.handleSharedBoxChangeCallback(contentKey, isShared)}} isShared={isShared ? true : false} />;
    this.setComponentToStateArray();
    // this.setState({contentComponentArray: this.state.contentComponentArray});

    // also have to put content title to state in order to track content and update
    this.contents[newContentKey] = addJournalContent ? addJournalContent : "";
    this.contents[`${newContentKey}-isShared`] = isShared ? true : false;
  }

  setComponentToStateArray(){
    // transfor object to array in order to display
    this.state.contentComponentArray=[];
    Object.keys(this.contentIndexs).map((element)=>{
      this.state.contentComponentArray.push(this.contentIndexs[element]);
    });
    this.setState({contentComponentArray: this.state.contentComponentArray});
  }

  handleSubmit(event, isUpdate){
    event.preventDefault();
    // console.log("submit: contents:",this.contents);
    console.log("journal date:",this.state.journalDate);
    // Assemble a journal object for save
    let readingIdArray = [];
    Object.keys(this.readings).map((key)=>{readingIdArray.push(key)});
    let journal = Object.assign({_id:this.jounalId, date: new Date(this.state.journalDate), ping_pong_state: this.state.pingPongState, user_id: this.userId, readingIds: readingIdArray}, this.contents);

    if(isUpdate){
      console.log("submit: journal:", journal);
      // console.log("oldContentKeys:", this.oldContentKeys);
      console.log("oldReadingKeys:", this.oldReadingIds);
      // Assemble two arrays for deletion contents and reading
      let deleteContents = [];
      let deleteReadingIds = [];
      // this.oldContentKeys.map((element)=>{if(!this.contents.hasOwnProperty(element)) deleteContents.push(element);});
      this.oldReadingIds.map((element)=>{if(!this.readings.hasOwnProperty(element)) deleteReadingIds.push(element);});
      // journal.deleteContents = deleteContents; //todo delete deleteContents and oldContentKeys
      journal.deleteReadingIds = deleteReadingIds;
      console.log("delete contents: ",deleteContents);
      console.log("delete reading ids : ",deleteReadingIds);
      DatabaseApi.updateJournal(journal).then((result)=>{
        this.props.history.push("/reading");
      });
    }else{
      DatabaseApi.createJournal(journal).then((result)=>{
        this.props.history.push("/reading");
      });
    }
    /**/
    this.setState({isWriting: true});
    console.log("submit: journal:", journal);
  }

  /* For ReadingSearchAndList callback */
  handleAttachReadingCallback(readingId){
    this.readings[readingId] = null;
    this.setState({isEmptyReading: false});
  }

  handleDetachAttachReadingCallback(readingId){
    delete this.readings[readingId];
    this.setState({isEmptyReading: Object.keys(this.readings).length===0});
  }

  handleCancel(){
    this.props.history.push("/reading");
  }

  handleDelete(){
    console.log("Delete journal!");
  }

  render(){
    return(
      <div className="addReadingDiv">
        {this.state.isWriting && <Loading text="Creating" />}

        <div className="titleDiv">{this.props.journalData ? "Update journal for readings" : "Add a new journal for readings"}</div>
        <form className="form-horizontal" onSubmit={(event) => {this.handleSubmit(event, this.props.journalData);}}>

          <div className="text-right bottom-btn-div">
            {(!this.props.journalData || this.userId == this.journalUserId) &&
              <button type="submit" className="btn btn-info loginButton" disabled={this.state.isWriting || !(this.state.journalDate.length>0) || !(this.state.isDateCorrect) || this.state.isEmptyReading}>{this.props.journalData ? "Update" : "Submit"}</button>
            }
            {(this.props.journalData && this.userId == this.journalUserId) &&
              <button onClick={()=>{this.handleDelete()}} type="button" className="btn btn-danger loginButton">Delete</button>
            }
            <button onClick={()=>{this.handleCancel()}} type="button" className="btn btn-nomal loginButton">Cancel</button>
          </div>

            <div className="form-group row form-div">
              <label htmlFor="journalDate" className="col-sm-1 col-form-label">Date</label>
              <div className="col-sm-4">

                <input className={this.state.isDateCorrect?"form-control":"form-control form-control-warning"} type="text" placeholder="mm/dd/yyyy" id="journalDate" value={this.state.journalDate} onChange={(event)=>{this.handleChange(event, "journalDate")}} />
                {!this.state.isDateCorrect && <span className="glyphicon glyphicon-warning-sign form-control-feedback form-control-warning-span"></span>}

              </div>
              <label htmlFor="pingPongState" className="col-sm-3 col-form-label">Ping Pong State</label>
              <div className="col-sm-4">
                <select value={this.state.pingPongState} id="pingPongState" className="form-control" onChange={(event)=>{this.handleChange(event,"pingPongState");}}>
                  <option value="Neutral">Neutral</option>
                  <option value="Reading/Call">Reading/Call</option>
                  <option value="Impact absorbing">Impact absorbing</option>
                  <option value="Impact complete">Impact complete</option>
                  <option value="Response">Response</option>
                  <option value="Outward moving">Outward moving</option>
                  <option value="Outward complete">Outward complete</option>
                </select>
              </div>
            </div>

            {/* New content goes here */}
            {this.state.contentComponentArray}

            {/* Add content button and drop list */}
            <div className="row addJournalContentDiv">

              <div onClick={()=>{this.handleAddContentClick();}} className="addJournalContentBtnDiv col-sm-6"><i className="fa fa-plus-square" /> Add one content for your journal</div>
              <div className="col-sm-6">
                <select  className="form-control" onChange={(event)=>{this.handleChange(event,"addJournalContent");}}>
                  <option value="overview">Overview</option>
                  <option value="internal_environment">Internal Environment</option>
                  <option value="relational_environment">Relational Environment</option>
                  <option value="physical_environment">Physical Environment</option>
                  <option value="creative_vector">Creative Vector</option>
                  <option value="synchronicities">Synchronicities</option>
                  <option value="dreams">Dreams</option>
                  <option value="other">Other</option>
                </select>

              </div>

            </div>


            {/*  Start reading search function */}
            <ReadingSearchAndList readings={this.props.journalData ? this.readings : null} attachReadingCallback={(readingId)=>{this.handleAttachReadingCallback(readingId);}} detachReadingCallback={(readingId)=>{this.handleDetachAttachReadingCallback(readingId);}} />

              <div className="text-right bottom-btn-div">
                {(!this.props.journalData || this.userId == this.journalUserId) &&
                  <button type="submit" className="btn btn-info loginButton" disabled={this.state.isWriting || !(this.state.journalDate.length>0) || !(this.state.isDateCorrect) || this.state.isEmptyReading}>{this.props.journalData ? "Update" : "Submit"}</button>
                }
                {(this.props.journalData && this.userId == this.journalUserId) &&
                  <button onClick={()=>{this.handleDelete()}} type="button" className="btn btn-danger loginButton">Delete</button>
                }
                <button onClick={()=>{this.handleCancel()}} type="button" className="btn btn-nomal loginButton">Cancel</button>
              </div>
          </form>

        </div>
    );
  }


}
AddJournal.propTypes={
  journalData: PropTypes.object,
  history: PropTypes.object
};
export default AddJournal;
