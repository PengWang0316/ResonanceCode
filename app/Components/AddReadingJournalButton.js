import React from 'react';
import { Link } from 'react-router-dom';
import jQuery from 'jquery';

import styles from '../styles/AddReadingJournalButton.module.css';

const handleAddBtnClick = () => {
  const isShowAddBtns = jQuery('#addReadingBtn').css('opacity') !== '0';
  const opacity = isShowAddBtns ? '0' : '1';
  const readingBottom = isShowAddBtns ? '25px' : '85px';
  const journalBottom = isShowAddBtns ? '25px' : '145px';
  jQuery('#addReadingBtn').css({ opacity, bottom: readingBottom });
  jQuery('#addJournalBtn').css({ opacity, bottom: journalBottom });
};

// $ will use jQuery in the index.html.
const handleAddReadingButtonClick = () => $('#addReadingModal').modal('toggle');

const AddReadingJurnalButton = () => (
  <div role="button" tabIndex="-3" onClick={handleAddBtnClick}>
    {/* this $('#addReadingModal').modal('toggle') will use jQuery from index.html */}
    <div role="button" tabIndex="-4" onClick={handleAddReadingButtonClick} id="addReadingBtn" className={`${styles.addBtn} ${styles.addReadingBtn} text-center`}>Reading</div>
    <Link to={{ pathname: '/addJournal' }}><div id="addJournalBtn" className={`${styles.addBtn} ${styles.addJournalBtn} text-center`}>Journal</div></Link>
    <div className={`${styles.addBtn}`}><i className="fa fa-plus" /></div>
  </div>
);

export default AddReadingJurnalButton;
