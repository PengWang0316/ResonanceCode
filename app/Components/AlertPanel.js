import React from 'react';

import styles from '../styles/AlertPanel.module.css';

const AlertPanel = ({
  id = '', type = 'success', children = '', style = {}
}) => (
  <div id={id} className={`alert alert-${type} mt-3 ${styles.alertPanel}`} style={style} role="alert">
    {children}
  </div>
);
export default AlertPanel;
