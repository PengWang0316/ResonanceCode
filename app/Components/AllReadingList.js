import React from 'react';
import { connect } from 'react-redux';

import Pagination from './SharedComponents/Pagination';
import LoadingAnimation from './SharedComponents/LoadingAnimation';
import { fetchAllReadingList as fetchAllReading } from '../actions/ReadingActions';
import { NUMBER_OF_READING_PER_PAGE } from '../config';
import ReadingListRow from './ReadingListRow';

/* Show the all reading in the list */
export const AllReadingList = ({
  allReadingList, readingsAmount, fetchAllReadingList, handleClick, isLoading
}) => (
  <div className="modal fade" id="readingListModal" tabIndex="-1" role="dialog" aria-labelledby="readingListModalLabel" aria-hidden="true">
    <div className="modal-dialog" role="document">
      <div className="modal-content">
        <div className="modal-header">
          <h5 className="modal-title" id="readingListModalLabel">Select a reading for this journal</h5>
          <button type="button" className="close" data-dismiss="modal" aria-label="Close">
            <span aria-hidden="true">&times;</span>
          </button>
        </div>
        <div className="modal-body">
          <LoadingAnimation />
          {!isLoading &&
            <div className="mb-3">
              {allReadingList.map(reading => <ReadingListRow key={`${reading.name}_${reading.date}`} reading={reading} handleClick={handleClick} />)}
            </div>
          }

          {/* Pagination componentWillMount
            * Keeping readingsAmount, numberPerpage, and fetchContent out of AllReadingListPagination component can help it raising reusability.
          */}
          {readingsAmount !== null && <Pagination
            amount={readingsAmount}
            fetchContent={fetchAllReadingList}
            numberPerpage={NUMBER_OF_READING_PER_PAGE}
          />}

        </div>
        <div className="modal-footer">
          <button type="button" className="btn btn-secondary" data-dismiss="modal">Close</button>
        </div>
      </div>
    </div>
  </div>
);
const mapStateToProps = state => ({
  allReadingList: state.allReadingList,
  readingsAmount: state.readingsAmount,
  isLoading: state.isLoading
});
const mapDispatchToProps = dispatch => ({
  fetchAllReadingList: pageNumber => dispatch(fetchAllReading(pageNumber))
});
// export AllReadingList;
export default connect(mapStateToProps, mapDispatchToProps)(AllReadingList);
