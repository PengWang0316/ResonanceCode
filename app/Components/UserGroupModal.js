import React, { Component } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';

import UserPicker from './UserPicker';
import { fetchUsersAmount, fetchAllUserList, updateUserGroup } from '../actions/UserActions';

/** The form that will be used to create and edit user group information. */
export class UserGroupModal extends Component {
  static propTypes = {
    userGroup: PropTypes.array,
    groupName: PropTypes.string.isRequired,
    isUpdate: PropTypes.bool.isRequired,
    users: PropTypes.array,
    usersAmount: PropTypes.number,
    user: PropTypes.object.isRequired,
    fetchUsersAmount: PropTypes.func.isRequired,
    fetchAllUserList: PropTypes.func.isRequired,
    updateUserGroup: PropTypes.func.isRequired
  };

  static defaultProps = {
    users: [],
    usersAmount: 0,
    userGroup: null
  };

  state = { userList: {}, groupName: '', isUpdate: false };

  /** Fetching the users and userAmount for the component.
    * @param {object} props is an object contains vaules for the component.
    * @return {null} No return.
  */
  componentWillMount() {
    this.props.fetchUsersAmount();
    this.props.fetchAllUserList(0);
  }

  /** When receive a new userGroup from props, save it to state.
    * @param {object} nextProps is an object that contains props' value.
    * @return {null} No return.
  */
  componentWillReceiveProps({ userGroup, groupName, isUpdate }) {
    this.setState({ userList: {}, groupName: '', isUpdate });
    if ((this.props.userGroup !== userGroup) &&
    userGroup) {
      const userList = {};
      userGroup.forEach(share => { userList[share.id] = share; });
      this.setState({ userList, groupName, isUpdate });
      // this.isUpdate = true;
    }
  }

  /** Updating the state when a user changes group name input.
    * @param {object} event is an object that repersents the element the user is reacting with.
    * @return {null} No return.
  */
  handleInputChange = ({ target }) => this.setState({ [target.id]: target.value });

  /** Adding the user to the state when a user click a user's name from user name list. The html element includes a img or i tag and a text element.
    * @param {object} target is the html element a user is clicking.
    * @return {null} No return.
  */
  handleAddUserCallback = ({ target }) => {
    let photo = null;
    let userId = null;
    if (target.childNodes[0]) { // If the target does not have a child node, a user is clicking on the avatar photo.
      photo = target.childNodes[0].getAttribute('src') ? target.childNodes[0].getAttribute('src') : '';
      userId = target.getAttribute('userid');
    } else { // Using a different approach to get information when a user click on the avater photo.
      photo = target.getAttribute('src') ? target.getAttribute('src') : '';
      userId = target.parentElement.getAttribute('userid');
    }
    this.setState({
      userList: Object.assign({
        [userId]: {
          id: userId,
          displayName: target.innerText,
          photo
        }
      }, this.state.userList)
    });
  };

  /** Add all user in the group to shareList.
    * @param {object} event is an object that repersents the element a user is clicking.
    * @return {null} No return.
  */
  handleGroupClickCallback = ({ target }) => {
    const addUserList = {};
    this.props.user.settings.userGroups[target.id]
      .forEach(({ id, displayName, photo }) => {
        addUserList[id] = {
          id,
          displayName,
          photo,
          sharedDate: new Date()
        };
      });
    this.setState({ userList: { ...addUserList, ...this.state.userList } });
  };

  /** Removing the user from state when the user click remove button.
    * @param {object} target is the html element a user is clicking.
    * @return {null} No return.
  */
  handleRemoveUserCallback = ({ target }) => {
    const newUserList = Object.assign({}, this.state.userList);
    delete newUserList[target.getAttribute('userid')];
    this.setState({ userList: newUserList });
  };

  /** Saving the sharing information back to the journal.
    * @return {null} No return.
  */
  handleSave = () => {
    this.props.updateUserGroup({ // Do have to go through Redux since the method has to update user state in Redux.
      isUpdate: this.state.isUpdate,
      newGroupName: this.state.groupName,
      oldGroupName: this.props.groupName, // If the component has oldGroupName and it is different with newGroupName, we have to remove the old data.
      userList: Object.keys(this.state.userList).map(key => this.state.userList[key])
    });
    $('#userGroupModal').modal('toggle'); // $ will use the jQuery libaray from index.html.
  };

  /** When a user clicks cancel button, reset state for next use.
    * @return {null} No return.
  */
  // handleCancelClick = () => {
  //   this.isUpdate = false; // Reset the update indicator.
  //   this.setState({ userList: {}, groupName: '' }); // also reset the states for next use.
  // };


  /** Rendering the jsx for the component.
    * @return {jsx} Return jsx for the component.
  */
  render() {
    return (
      <div className="modal fade" id="userGroupModal" tabIndex="-1" role="dialog" aria-labelledby="userGroupModalLabel" aria-hidden="true">
        <div className="modal-dialog" role="document">
          <div className="modal-content">
            <div className="modal-header">
              <h5 className="modal-title" id="userGroupModalLabel">User group</h5>
              <button type="button" className="close" data-dismiss="modal" aria-label="Close">
                <span aria-hidden="true">&times;</span>
              </button>
            </div>
            <div className="modal-body">
              <div className="form-group row font-weight-bold">
                <label htmlFor="inputGroupName" className="col-sm-3 col-form-label">Group name</label>
                <div className="col-sm-9">
                  <input type="text" className="form-control" id="groupName" placeholder="The name of user group..." value={this.state.groupName} onChange={this.handleInputChange} />
                </div>
              </div>
              <UserPicker
                userList={this.state.userList}
                handleRemoveUser={this.handleRemoveUserCallback}
                chooseTextTitle="Pick users you want to put in this group"
                users={this.props.users}
                usersAmount={this.props.usersAmount}
                user={this.props.user}
                removeText="Move out this user to the group"
                handleAddUser={this.handleAddUserCallback}
                fetchAllUserList={fetchAllUserList}
                handleGroupClick={this.handleGroupClickCallback}
              />
            </div>
            <div className="modal-footer">
              <button type="button" className="btn btn-secondary" data-dismiss="modal">Close</button>
              <button type="button" className="btn btn-primary" disabled={this.state.groupName === '' || Object.keys(this.state.userList).length === 0} onClick={this.handleSave}>Save changes</button>
            </div>
          </div>
        </div>
      </div>
    );
  }
}
/* istanbul ignore next */
const mapStateToProps = state => ({
  users: state.users,
  usersAmount: state.usersAmount,
  user: state.user
});
/* istanbul ignore next */
const mapDispatchToProps = dispatch => ({
  fetchUsersAmount: _ => dispatch(fetchUsersAmount()),
  fetchAllUserList: _ => dispatch(fetchAllUserList()),
  updateUserGroup: params => dispatch(updateUserGroup(params))
});
export default connect(mapStateToProps, mapDispatchToProps)(UserGroupModal);
