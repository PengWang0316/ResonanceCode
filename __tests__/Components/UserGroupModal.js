import React from 'react';
import renderer from 'react-test-renderer';
import { shallow } from 'enzyme';
import sinon from 'sinon';

import { UserGroupModal } from '../../app/Components/UserGroupModal';

jest.mock('../../app/Components/UserPicker', () => 'UserPicker');

describe('UserGroupModal test', () => {
  const defaultProps = {
    userGroup: [{ id: 'id1' }, { id: 'id2' }],
    groupName: '',
    isUpdate: false,
    user: { settings: { userGroups: { id1: [{ id: 'idA', displayName: 'displayNameA', photo: 'photoA' }] } } },
    fetchUsersAmount: jest.fn(),
    fetchAllUserList: jest.fn(),
    updateUserGroup: jest.fn()
  };
  const getShallowComponent = (props = defaultProps) => shallow(<UserGroupModal {...props} />);

  test('componentWillMount and initial states', () => {
    const component = getShallowComponent();
    expect(defaultProps.fetchUsersAmount).toHaveBeenCalledTimes(1);
    expect(defaultProps.fetchAllUserList).toHaveBeenLastCalledWith(0);
    expect(component.state('userList')).toEqual({});
    expect(component.state('userList')).toEqual({});
    expect(component.state('groupName')).toBe('');
    expect(component.state('isUpdate')).toBeFalsy();
  });

  test('UserGroupModal snapshot', () => expect(renderer.create(<UserGroupModal {...defaultProps} />).toJSON()).toMatchSnapshot());

  test('componentWillReceiveProps', () => {
    const component = getShallowComponent();
    component.setState({ userList: {}, groupName: 'aaa', isUpdate: false });
    component.setProps({ userGroup: [{ id: 'idA' }, { id: 'idB' }], groupName: 'newGroupName', isUpdate: true });
    expect(component.state('userList')).toEqual({ idA: { id: 'idA' }, idB: { id: 'idB' } });
    expect(component.state('groupName')).toBe('newGroupName');
    expect(component.state('isUpdate')).toBeTruthy();
  });

  test('handleInputChange', () => {
    const component = getShallowComponent();
    component.find('input').simulate('change', { target: { id: 'testId', value: 'testValue' } });
    expect(component.state('testId')).toBe('testValue');
  });

  test('handleAddUserCallback has childNodes', () => {
    const component = getShallowComponent();
    component.setState({ userList: { oldId: { id: 'oldId', displayName: 'oldName', photo: 'oldPhoto' } } });
    const mockFuncA = jest.fn().mockReturnValue('photoSrc');
    const mockFuncB = jest.fn().mockReturnValue('userId');
    component.find('UserPicker').prop('handleAddUser')({ target: { getAttribute: mockFuncB, childNodes: [{ getAttribute: mockFuncA }], innerText: 'displayNameText' } });
    expect(mockFuncA).toHaveBeenLastCalledWith('src');
    expect(mockFuncB).toHaveBeenLastCalledWith('userid');
    expect(component.state('userList')).toEqual({ userId: { id: 'userId', displayName: 'displayNameText', photo: 'photoSrc' }, oldId: { id: 'oldId', displayName: 'oldName', photo: 'oldPhoto' } });
  });

  test('handleAddUserCallback without childNodes', () => {
    const component = getShallowComponent();
    component.setState({ userList: { oldId: { id: 'oldId', displayName: 'oldName', photo: 'oldPhoto' } } });
    const mockFuncA = jest.fn().mockReturnValue('photoSrc');
    const mockFuncB = jest.fn().mockReturnValue('userId');
    component.find('UserPicker').prop('handleAddUser')({
      target: {
        getAttribute: mockFuncA, parentElement: { getAttribute: mockFuncB }, childNodes: [null], innerText: 'displayNameText'
      }
    });
    expect(mockFuncA).toHaveBeenLastCalledWith('src');
    expect(mockFuncB).toHaveBeenLastCalledWith('userid');
    expect(component.state('userList')).toEqual({ userId: { id: 'userId', displayName: 'displayNameText', photo: 'photoSrc' }, oldId: { id: 'oldId', displayName: 'oldName', photo: 'oldPhoto' } });
  });

  test('handleGroupClickCallback', () => {
    sinon.useFakeTimers(new Date(2018, 0, 1));
    const component = getShallowComponent();
    component.setState({ userList: { oldId: { id: 'oldId', displayName: 'oldName', photo: 'oldPhoto' } } });
    component.find('UserPicker').prop('handleGroupClick')({ target: { id: 'id1' } });
    expect(component.state('userList')).toEqual({
      idA: {
        id: 'idA', displayName: 'displayNameA', photo: 'photoA', sharedDate: new Date()
      },
      oldId: { id: 'oldId', displayName: 'oldName', photo: 'oldPhoto' }
    });
  });

  test('handleRemoveUserCallback', () => {
    const component = getShallowComponent();
    const mockFunc = jest.fn().mockReturnValue('oldId');
    component.setState({ userList: { oldId: { id: 'oldId', displayName: 'oldName', photo: 'oldPhoto' } } });
    component.find('UserPicker').prop('handleRemoveUser')({ target: { getAttribute: mockFunc } });
    expect(component.state('userList')).toEqual({});
  });

  test('handleSave', () => {
    const component = getShallowComponent();
    component.setProps({ groupName: 'oldGroupName' });
    component.setState({
      userList: { oldId: { id: 'oldId', displayName: 'oldName', photo: 'oldPhoto' } }, isUpdate: true, groupName: 'groupName'
    });
    const mockFunc = jest.fn();
    window.$ = jest.fn().mockReturnValue({ modal: mockFunc });
    component.find('button').at(2).simulate('click');
    expect(defaultProps.updateUserGroup).toHaveBeenLastCalledWith({
      isUpdate: true, newGroupName: 'groupName', oldGroupName: 'oldGroupName', userList: [{ id: 'oldId', displayName: 'oldName', photo: 'oldPhoto' }]
    });
    expect(window.$).toHaveBeenLastCalledWith('#userGroupModal');
    expect(mockFunc).toHaveBeenLastCalledWith('toggle');
  });

  test('button disabled', () => {
    const component = getShallowComponent();
    component.setState({ groupName: '' });
    expect(component.find('button').at(2).prop('disabled')).toBe(true);
    component.setState({ groupName: 'aa', userList: {} });
    expect(component.find('button').at(2).prop('disabled')).toBe(true);
    component.setState({ groupName: 'aa', userList: { id: 'a' } });
    expect(component.find('button').at(2).prop('disabled')).toBe(false);
  });
});
