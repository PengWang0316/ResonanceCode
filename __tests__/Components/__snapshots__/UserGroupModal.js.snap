// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserGroupModal test UserGroupModal snapshot 1`] = `
<div
  aria-hidden="true"
  aria-labelledby="userGroupModalLabel"
  className="modal fade"
  id="userGroupModal"
  role="dialog"
  tabIndex="-1"
>
  <div
    className="modal-dialog"
    role="document"
  >
    <div
      className="modal-content"
    >
      <div
        className="modal-header"
      >
        <h5
          className="modal-title"
          id="userGroupModalLabel"
        >
          User group
        </h5>
        <button
          aria-label="Close"
          className="close"
          data-dismiss="modal"
          type="button"
        >
          <span
            aria-hidden="true"
          >
            Ã—
          </span>
        </button>
      </div>
      <div
        className="modal-body"
      >
        <div
          className="form-group row font-weight-bold"
        >
          <label
            className="col-sm-3 col-form-label"
            htmlFor="inputGroupName"
          >
            Group name
          </label>
          <div
            className="col-sm-9"
          >
            <input
              className="form-control"
              id="groupName"
              onChange={[Function]}
              placeholder="The name of user group..."
              type="text"
              value=""
            />
          </div>
        </div>
        <UserPicker
          chooseTextTitle="Pick users you want to put in this group"
          fetchAllUserList={[Function]}
          handleAddUser={[Function]}
          handleGroupClick={[Function]}
          handleRemoveUser={[Function]}
          removeText="Move out this user to the group"
          user={
            Object {
              "settings": Object {
                "userGroups": Object {
                  "id1": Array [
                    Object {
                      "displayName": "displayNameA",
                      "id": "idA",
                      "photo": "photoA",
                    },
                  ],
                },
              },
            }
          }
          userList={Object {}}
          users={Array []}
          usersAmount={0}
        />
      </div>
      <div
        className="modal-footer"
      >
        <button
          className="btn btn-secondary"
          data-dismiss="modal"
          type="button"
        >
          Close
        </button>
        <button
          className="btn btn-primary"
          disabled={true}
          onClick={[Function]}
          type="button"
        >
          Save changes
        </button>
      </div>
    </div>
  </div>
</div>
`;
